File: node/get-broadcast-opens-clicks.js
const fetch = require("node-fetch");
const credentials = require("./credentials.json");
const BASE_URL = 'https://api.aweber.com/1.0/';
const client_id = credentials['clientId'];
const client_secret = credentials['clientSecret'];
const accessToken = credentials['accessToken'];

const qs = require("querystring");

/**
 * Get all of the entries for a collection
 *
 * @param string $accessToken Access token to pass in as an authorization header
 * @param string $url Full url to make the request
 * @return array Every entry in the collection
 */

async function getCollection(accessToken,url) {
    let res;
    const collection = [];
    console.log({accessToken})
    while(url) {
        res = await fetch(url,{
            headers:{
                "Authorization":`Bearer ${accessToken}`
            }
        });
        let page = await res.json();
        console.log("got page",{page})
        collection.push(...page.entries);
        url = page.next_collection_link;
    }
    return collection;
}

/**
 * Get a URL, retrying on rate limit errors
 *
 * @param Client client guzzle client instance
 * @param string url URL to get
 * @return \Psr\Http\Message\ResponseInterface
 */
const MAX_RETRIES = 5;
const wait = (ms) => new Promise(resolve => setTimeout(resolve,ms));
// throw an error if non success response
async function request(url,options) {
    if (options == null) options = {};
    if (options.credentials == null) options.credentials = 'same-origin';
    return fetch(url, options).then(function(response) {
        if (response.status >= 200 && response.status < 300) {
            return Promise.resolve(response);
        } else {
            var error = new Error(response.statusText || response.status);
            error.response = response;
            return Promise.reject(error);
        }
    });
}
async function getWithRetry(accessToken, url) {
    for (let i = 0;i< MAX_RETRIES; i++ ) {
        try {
            const response = await request(url,{
                'headers' : {
                    'Authorization': 'Bearer ' + accessToken
                }
            });
            return response;

        } catch(e) {
            // Only retry on a 403 (forbidden) status code with a rate limit error
            if (!e.response || e.response.statusCode !== 403) {
                throw e;
            }

            const body = await e.response.json();

            if (!body['error']['message'].match(/rate limit/i)) {
                throw new Error(body['error']['message']);
            }
            console.log("Request was rate limited\n");
            if (i < MAX_RETRIES) {
                // Wait longer between every attempt
                await wait(2 * i * 1000);
                console.log(`Retry ${i}...\n`);
            }
            continue;
        }
    }

    throw new Error ("Giving up after " + MAX_RETRIES + " tries");
}

(async () => {

    // get all the accounts entries
    const accounts = await getCollection(accessToken, BASE_URL + 'accounts');
    console.log({accounts})
    const accountUrl = accounts[0]['self_link'];

    // get all the list entries for the first account
    const listsUrl = accounts[0]['lists_collection_link'];
    const lists = await getCollection( accessToken, listsUrl);

    // get a sent broadcast
    const params =  {
        'ws.op' : 'find',
        'campaign_type':  'b'
    }
    const campaignsUrl = lists[0]['campaigns_collection_link'];
    const broadcastsUrl = campaignsUrl + '?' + qs.stringify(params);
    console.log(broadcastsUrl)
    const sentBroadcasts = await getCollection( accessToken, broadcastsUrl);
    const broadcast = sentBroadcasts[0];
    console.log ('Broadcast: ');
    console.log(broadcast);

    // A broadcast is the receipt of sending email messages to a list.
    // Messages are each email sent to a subscriber; they
    // can be opened and clicked by each subscriber multiple times.
    const messagesUrl = broadcast['messages_collection_link'];
    const messages = await getCollection(accessToken, messagesUrl);

    // mapping of subscriber url to email address
    const subscriberCache = {};

    console.log ("Opens for broadcast:");
    for (let message of messages) {
        if (message['total_opens'] > 0) {
            // You could also paginate the opens collection of each message,
            // but the open count and the last_opened timestamp are in the message entry
            // so unless we need exact times of each non-unique open, we use the message entry.
            const openSubLink = message['subscriber_link'];
            const openSub = await getWithRetry( accessToken, openSubLink);
            const openSubBody = await openSub.json();
            console.log (`     ${message['last_opened']}: ${openSubBody['email']}`);
            // First time looking up a subscriber; save them for next time
            subscriberCache[openSubLink] = openSubBody['email'];
        }
    }

    const links = await getCollection(accessToken, broadcast['links_collection_link']);

    console.log ("Clicks for broadcast:");
    for (let link of links) {
        console.log( "{link['url']}\n");
        const clicksUrl = link['clicks_collection_link'];
        const clicks = await getCollection( accessToken, clicksUrl);
        for (let click of clicks) {
            const clickSubLink = click['subscriber_link'];
            const cachedSubscriber = subscriberCache[clickSubLink];
            let email;
            if (cachedSubscriber) {
                email = cachedSubscriber['email'];
            } else {
                const clickSub = await getWithRetry( accessToken, clickSubLink);
                const clickSubBody = await clickSub.json();
                email = clickSubBody['email'];
            }
            console.log(`    ${click['event_time']}: ${email}`);
        }
    }
})();

File: ruby/get_broadcast_opens_clicks.rb
# frozen_string_literal: true

require 'faraday'
require 'json'
require 'time'



MAX_RETRIES = 5
BASE_URL = 'https://api.aweber.com/1.0/'
credentials = JSON.parse(File.read('./credentials.json'))
puts credentials
conn = Faraday.new(
  headers: {
    Authorization: "Bearer #{credentials['access_token']}"
  }
) do |f|
  f.response :raise_error
end

def get_collection(conn, url)
  collection = []
  while url
    res = conn.get(url)
    page = JSON.parse(res.body)
    collection.push(*page['entries'])
    url = page['next_collection_link']
  end
  collection
end

def get_with_retry(url,credentials)
  Faraday.new(
    headers: {
      Authorization: "Bearer #{credentials['access_token']}"
    }
  ) do |f|
    f.request :retry, { max: MAX_RETRIES, interval: 2 }
  end.get(url)
end
accounts = get_collection(conn, "#{BASE_URL}accounts")

lists_url = accounts[0]['lists_collection_link']
lists = get_collection(conn, lists_url)

broadcasts_url = lists[0]['sent_broadcasts_link']
puts broadcasts_url

sent_broadcasts = get_collection(conn, broadcasts_url)
broadcast_response = get_with_retry(sent_broadcasts[0]['self_link'], credentials)
broadcast = JSON.parse(broadcast_response.body)
puts "Broadcast:"
puts broadcast

puts "Opens for broadcast:"
opens_url = broadcast['opens_collection_link']
opens = get_collection(conn, opens_url)

opens.each do |open|
  puts "    [#{open['event_time']}]: #{open['email']}"
end

puts "Clicks for broadcast:"
clicks_url = broadcast['clicks_collection_link']
clicks = get_collection(conn, clicks_url)

clicks.each do |open|
  puts "    [#{open['event_time']}]: #{open['email']}"
end
